src/backends/r1cs/mod.rs

```diff
@@ -0,0 +1,489 @@
+/// a and b are represented by CellVar.
+/// The constant f_c is represented by the constant field, will always multiply with the variable at index 0 which is always 1.
+#[derive(Clone, Debug)]
+pub struct LinearCombination {
+    pub terms: HashMap<CellVar, Fr>,
+    pub constant: Fr,
+}
+
+impl LinearCombination {
+    /// Evaluate the linear combination with the given witness.
+    pub fn evaluate(&self, witness: &[Fr]) -> Fr {
+        let mut sum = Fr::zero();
+
+        for (var, factor) in &self.terms {
+            sum += *witness.get(var.index).unwrap() * factor;
+        }
+
+        sum += &self.constant;
+
+        sum
+    }
+
+    /// Create a linear combination to represent constant one.
+    pub fn one() -> Self {
+        LinearCombination {
+            terms: HashMap::new(),
+            constant: Fr::from(1),
+        }
+    }
+
+    /// Create a linear combination to represent constant zero.
+    pub fn zero() -> Self {
+        LinearCombination {
+            terms: HashMap::new(),
+            constant: Fr::from(0),
+        }
+    }
+
+    /// Create a linear combination from a list of vars
+    pub fn from_vars(vars: Vec<CellVar>) -> Self {
+        let terms = vars.into_iter().map(|var| (var, Fr::from(1))).collect();
+        LinearCombination {
+            terms,
+            constant: Fr::from(0),
+        }
+    }
+
+    /// Create a linear combination from a constant.
+    pub fn from_const(cst: Fr) -> Self {
+        LinearCombination {
+            terms: HashMap::new(),
+            constant: cst,
+        }
+    }
+}
+
+/// an R1CS constraint
+/// Each constraint comprises of 3 linear combinations from 3 matrices.
+/// It represents a constraint in math: a * b = c.
+#[derive(Clone, Debug)]
+pub struct Constraint {
+    pub a: LinearCombination,
+    pub b: LinearCombination,
+    pub c: LinearCombination,
+}
+
+impl Constraint {
+    /// Convert the 3 linear combinations to an array.
+    pub fn as_array(&self) -> [&LinearCombination; 3] {
+        [&self.a, &self.b, &self.c]
+    }
+}
+
+
+/// R1CS backend with bls12_381 field.
+#[derive(Clone)]
+pub struct R1csBls12_381 {
+    /// Constraints in the r1cs.
+    constraints: Vec<Constraint>,
+    witness_vars: Vec<Value<R1csBls12_381>>,
+    debug_info: Vec<DebugInfo>,
+    /// Record the public inputs for reordering the witness vector
+    public_inputs: Vec<CellVar>,
+    /// Record the public outputs for reordering the witness vector
+    public_outputs: Vec<CellVar>,
+    finalized: bool,
+}
+
+impl R1csBls12_381 {
+    pub fn new() -> Self {
+        Self {
+            constraints: Vec::new(),
+            witness_vars: Vec::new(),
+            debug_info: Vec::new(),
+            public_inputs: Vec::new(),
+            public_outputs: Vec::new(),
+            finalized: false,
+        }
+    }
+
+    // todo: how can we create this r1cs backend with different associated field types, but still using the same backend implementation? using macro?
```

Comment:
I don't think you need to use macro, you could just have a `R1CS` struct generic on the prime, and then instantiate it in two different backends.